sycl/test/check_device_code/device_global_static.cpp

```diff
@@ -0,0 +1,29 @@
+// Tests that the llvm.compiler.used symbol, which is used to implement static
```

Comment:
Why `sycl/test` instead of `sycl-post-link` test?

Reply:
What is the latest stage it can be removed? Why can't we have test for that component? Is the "mess" final, or do we expect pipelines to be unified?

Reply:
`llvm.used` should survive to final linking, `llvm.compiler.used` is okay for intermediary tools to drop: https://llvm.org/docs/LangRef.html#the-llvm-used-global-variable

Reply:
> Do we need a test when programmer uses `__attribute__((used))` on something to ensure that something isn't being dropped?

I think that's probably a good idea but perhaps out of the scope of this patch.

Reply:
This PR is still too complex to me, and I can't comprehend how everything comes to play together. Maybe someone else from @intel/llvm-reviewers-runtime would be a better reviewer here.

Reply:
I've updated the description to more accurately describe what's happening. Let me know if you need further explanation and I'd be happy to provide.

Reply:
Much better now, everything but this test makes sense to me. How do we know that "llvm.compiler.used" string is ever present? What if bitcode format has a dedicated `builtin_id` for it? Can we use AOT instead to make sure that nothing unsupported reaches backend compiler?

Reply:
We are testing for that [here](https://github.com/intel/llvm/blob/sycl/clang/test/CodeGenSYCL/device_global.cpp#L484). Even if a triple doesn't use `llvm.compiler.used`, it's OK to make sure the binary has had the symbol removed once linking has happened. We are doing this as a full compilation pipeline since, as mentioned, the symbol gets removed at `sycl-post-link` for SPIR-V, and lowering for NVPTX and AMDGCN.

Reply:
Sorry, I don't see how this answers my question. My issue is with `strings`, not with the `FileCheck`'s checks.

Reply:
I think it's a byproduct. Why don't we want it? Because JIT/AOT would fail if it was present? We have a JIT test under e2e being added as part of this PR already. I'm not even sure one extra is necessary, but if so, I think just having an AOT compilation should suffice.

Reply:
See the latest commit [here](https://github.com/intel/llvm/pull/15224/commits/d3c9529acf8ac3a9943cae73cdf544d857dedb90). I don't think we need an extra JIT test for this as I am only changing behaviour in this PR for NVPTX and AMDGCN, which are both AOT.

Reply:
I still dislike it. @uditagarwal97 is developing an image compression feature, one potential implementation of it is that while `-fsycl-device-only` would keep producing LLVM bitcode, the final executable won't have it directly. The test is still way too fragile.

Reply:
I am testing that the file doesn't contain `llvm.compiler.used` in any section of the final binary. This could be a global symbol in plain text PTX or a symbol in compiler SPIR-V. I just need to make sure that the symbol (string) gets removed at some point in compilation. Do you have any suggestions on how I can improve this test?

Reply:
We are already testing for the global symbol removal at both LLVM lowering for NVPTX, AMDGCN and `sycl-post-link` for SPIRV. However since the SYCL TC uses different ways to remove this symbol, this test ensures that, while the components already work, the entire compilatino pipeline succeeds in removing this symbol.

Reply:
> I am testing that the file doesn't contain llvm.compiler.used in any section of the final binary.

Why is that important?

Reply:
I would consider it to be a bug if an executable contains orphaned symbols that are leftover from implementation specific hacks

Reply:
Does anything break if that happens?

Reply:
Yes

Reply:
In the case of using a compressed device image would a tool like `objdump` be able to find symbols in the compressed binary?

Reply:
Device image compression will be optional (https://github.com/intel/llvm/pull/15124) - unless you specify `--offload-compress` in CLI, by default, we won't compress images. So, image compression might not be an issue for this particular test.

> In the case of using a compressed device image would a tool like objdump be able to find symbols in the compressed binary

Nope. You'd have to retrieve the compressed device image from the executable and uncompress it using `zstd`, to find the symbols.

Reply:
> Device image compression will be optional (#15124) - unless you specify `--offload-compress` in CLI, by default, we won't compress images. So, image compression might not be an issue for this particular test.

If image compression is "opt-in" and we will not opt in in this test, then why is this an issue @aelovikov-intel ?

Reply:
Thanks for response @uditagarwal97 :)

Reply:
> If I just delete the test will you approve this PR? I don't think this discussion is going anywhere

Yes.